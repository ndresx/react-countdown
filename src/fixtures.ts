type TMockDateNow = {
  readonly now: jest.Mock;
  readonly timeDiff: number;
};

export function mockDateNow(): TMockDateNow {
  const timeDiff = 90110456;
  const now = jest.fn(() => 1482363367071);
  Date.now = now;

  return { now, timeDiff };
}

export const defaultStats = {
  total: 0,
  days: 0,
  hours: 0,
  minutes: 0,
  seconds: 0,
  milliseconds: 0,
  completed: false,
};
