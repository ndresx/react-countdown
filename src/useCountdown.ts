import { useState, useEffect, useRef } from 'react';

import CountdownJs, { CountdownProps, CountdownRenderProps, CountdownState } from './CountdownJs';

export interface UseCountdownProps extends Omit<CountdownProps, 'renderer'> {}

export interface UseCountdownResult extends CountdownRenderProps {}

export default function useCountdown(props: UseCountdownProps): UseCountdownResult {
  const [, setState] = useState<Partial<CountdownState>>({});
  const countdownObj = useRef<CountdownJs | null>(null);
  const firstRun = useRef(false);
  const key = useRef(props.key);

  if (!countdownObj.current || key.current !== props.key) {
    if (countdownObj.current) {
      key.current = props.key;
      firstRun.current = false;
    }

    countdownObj.current = new CountdownJs({ ...props }, (state, callback) => {
      setState(state);
      callback();
    });
  }

  const countdown = countdownObj.current as CountdownJs;

  useEffect(() => {
    countdown.init();
    return countdown.destroy;
  }, [countdown]);

  useEffect(() => {
    if (firstRun.current) {
      countdown.update(props);
    } else {
      firstRun.current = true;
    }
  }, [countdown, props]);

  return countdown.getRenderProps();
}
