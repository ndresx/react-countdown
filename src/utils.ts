export interface CountdownTimeDeltaOptions {
  readonly now?: () => number;
  readonly precision?: number;
  readonly controlled?: boolean;
}

export interface CountdownTimeDelta {
  readonly total: number;
  readonly days: number;
  readonly hours: number;
  readonly minutes: number;
  readonly seconds: number;
  readonly milliseconds: number;
  readonly completed: boolean;
}

export interface CountdownTimeDeltaFormatted {
  readonly days: string;
  readonly hours: string;
  readonly minutes: string;
  readonly seconds: string;
}

export interface CountdownTimeDeltaFormatOptions {
  readonly daysInHours?: boolean;
  readonly zeroPadLength?: number;
  readonly zeroPadDaysLength?: number;
}

/**
 * Pads a given string or number with zeros.
 *
 * @export
 * @param {number|string} value Value to zero-pad.
 * @param {number} [length=2] Amount of characters to pad.
 * @returns Left-padded number/string.
 */
export function zeroPad(value: number | string, length: number = 2): string {
  const strValue = String(value);
  if (length === 0) return strValue;
  return strValue.length >= length ? strValue : ('0'.repeat(length) + strValue).slice(length * -1);
}

export const timeDeltaFormatOptionsDefaults: CountdownTimeDeltaFormatOptions = {
  daysInHours: false,
  zeroPadLength: 2,
};

/**
 * Calculates the time difference between a given end date and the current date.
 *
 * @export
 * @param {Date|number|string} date Date or timestamp representation of the end date.
 * @param {Object} [{ now = Date.now, precision = 0, controlled = false }={}]
 *  {function} [date=Date.now] Alternative function for returning the current date.
 *  {number} [precision=0] The precision on a millisecond basis.
 *  {boolean} [controlled=false] Defines whether the calculated value is already provided as the time difference or not.
 * @param {number} [precision=0] The precision on a millisecond basis.
 * @param {boolean} [controlled=false] Defines whether the calculated value is already provided as the time difference or not.
 * @returns Object that includes details about the time difference.
 */
export function calcTimeDelta(
  date: Date | string | number,
  { now = Date.now, precision = 0, controlled = false }: CountdownTimeDeltaOptions = {}
): CountdownTimeDelta {
  let startTimestamp: number;

  if (typeof date === 'string') {
    startTimestamp = new Date(date).getTime();
  } else if (date instanceof Date) {
    startTimestamp = date.getTime();
  } else {
    startTimestamp = date;
  }

  const total = Math.round(
    parseFloat(
      (Math.max(0, controlled ? startTimestamp : startTimestamp - now()) / 1000).toFixed(
        Math.max(0, Math.min(20, precision))
      )
    ) * 1000
  );

  const seconds = total / 1000;

  return {
    total,
    days: Math.floor(seconds / (3600 * 24)),
    hours: Math.floor((seconds / 3600) % 24),
    minutes: Math.floor((seconds / 60) % 60),
    seconds: Math.floor(seconds % 60),
    milliseconds: Number(((seconds % 1) * 1000).toFixed()),
    completed: total <= 0,
  };
}

/**
 * Formats a given countdown time delta object
 *
 * @export
 * @param {CountdownTimeDelta} delta
 * @param {CountdownTimeDeltaFormatOptions} [options]
 * @returns {CountdownTimeDeltaFormatted} Formatted time delta object.
 */
export function formatTimeDelta(
  delta: CountdownTimeDelta,
  options?: CountdownTimeDeltaFormatOptions
): CountdownTimeDeltaFormatted {
  const { days, hours, minutes, seconds } = delta;
  const { daysInHours, zeroPadLength, zeroPadDaysLength = zeroPadLength } = {
    ...timeDeltaFormatOptionsDefaults,
    ...options,
  };
  const formattedHours = daysInHours
    ? zeroPad(hours + days * 24, zeroPadLength)
    : zeroPad(hours, Math.min(2, zeroPadLength));

  return {
    days: daysInHours ? '' : zeroPad(days, zeroPadDaysLength),
    hours: formattedHours,
    minutes: zeroPad(minutes, Math.min(2, zeroPadLength)),
    seconds: zeroPad(seconds, Math.min(2, zeroPadLength)),
  };
}
