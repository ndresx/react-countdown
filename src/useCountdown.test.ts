import { renderHook, act, cleanup } from '@testing-library/react-hooks';

const classSpies = {
  init: jest.fn(),
  update: jest.fn(),
  destroy: jest.fn(),
};

const CountdownJs = jest.requireActual('./CountdownJs').default;

jest.mock('./CountdownJs', () => {
  return jest.fn().mockImplementation((props, updater) => {
    const countdown = new CountdownJs(props, updater);
    Object.keys(classSpies).forEach(key => {
      const fn = countdown[key];
      countdown[key] = function(...args): void {
        fn(...args);
        classSpies[key](...args);
      };
    });
    return countdown;
  });
});

import useCountdown, { UseCountdownProps } from './useCountdown';
import { calcTimeDelta } from './utils';
import { mockDateNow, defaultStats } from './fixtures';

const { now, timeDiff } = mockDateNow();

describe('useCountdown', () => {
  jest.useFakeTimers();

  let countdownDate: number;
  const countdownMs = 10000;

  beforeEach(() => {
    Date.now = now;
    countdownDate = Date.now() + countdownMs;
  });

  const hookCallback = (props: UseCountdownProps) => useCountdown(props); // eslint-disable-line

  it('should use countdown', () => {
    const { result } = renderHook(hookCallback, { initialProps: { date: countdownDate } });
    expect(result.current).toMatchSnapshot();
  });

  it('should update props', () => {
    const { result, rerender } = renderHook(hookCallback, {
      initialProps: { date: countdownDate },
    });
    const prevDate = result.current.props.date;
    expect(prevDate).toBe(1482363377071);

    expect(classSpies.init).toHaveBeenCalledTimes(1);
    expect(classSpies.update).toHaveBeenCalledTimes(0);
    expect(classSpies.destroy).toHaveBeenCalledTimes(0);

    const nextProps = { date: countdownDate + timeDiff };
    rerender(nextProps);

    const nextDate = result.current.props.date;
    expect(nextDate).not.toEqual(prevDate);
    expect(nextDate).toBe(1482453487527);

    expect(classSpies.init).toHaveBeenCalledTimes(1);
    expect(classSpies.update).toHaveBeenCalledTimes(1);
    expect(classSpies.update).toHaveBeenLastCalledWith(nextProps);
    expect(classSpies.destroy).toHaveBeenCalledTimes(0);
  });

  it('should respect hook lifecycle', async () => {
    const { result, rerender } = renderHook(hookCallback, {
      initialProps: { date: countdownDate, intervalDelay: 1111 },
    });

    expect(classSpies.init).toHaveBeenCalledTimes(1);
    expect(classSpies.update).toHaveBeenCalledTimes(0);
    expect(classSpies.destroy).toHaveBeenCalledTimes(0);

    rerender({ date: countdownDate, intervalDelay: 1111 });

    act(() => {
      // Forward 6s in time
      now.mockReturnValue(countdownDate - 6000);
      jest.advanceTimersByTime(6000);
    });

    expect(classSpies.init).toHaveBeenCalledTimes(1);
    expect(classSpies.update).toHaveBeenCalledTimes(1);
    expect(classSpies.destroy).toHaveBeenCalledTimes(0);
    expect(result.current.total).toBe(6000);

    // Respect key-prop change and re-instantiate countdown
    rerender({ date: countdownDate, key: Math.random(), intervalDelay: 1111 });

    expect(classSpies.init).toHaveBeenCalledTimes(2);
    expect(classSpies.update).toHaveBeenCalledTimes(1);
    expect(classSpies.destroy).toHaveBeenCalledTimes(1);
    expect(result.current.total).toBe(6000);
  });

  [true, false].forEach(shouldUnmount => {
    it(`should update (unmount => ${shouldUnmount}) time total`, () => {
      const { result, unmount } = renderHook(hookCallback, {
        initialProps: { date: countdownDate, intervalDelay: 2222 },
      });

      act(() => {
        now.mockReturnValue(countdownDate - 6000);
        jest.advanceTimersByTime(6000);
      });

      expect(result.current.total).toBe(6000);

      if (shouldUnmount) unmount();

      act(() => {
        now.mockReturnValue(countdownDate - 3000);
        jest.advanceTimersByTime(3000);
      });

      expect(result.current.total).toBe(shouldUnmount ? 6000 : 3000);
    });
  });

  it('should trigger onTick and onComplete callbacks', () => {
    const onTick = jest.fn(stats => {
      expect(stats).toEqual(calcTimeDelta(countdownDate));
    });

    const onComplete = jest.fn(stats => {
      expect(stats.total).toEqual(0);
    });

    const { result } = renderHook(hookCallback, {
      initialProps: { date: countdownDate, onTick, onComplete },
    });

    act(() => {
      // Forward 6s in time
      now.mockReturnValue(countdownDate - 6000);
      jest.advanceTimersByTime(6000);
    });

    expect(onTick.mock.calls.length).toBe(6);
    expect(result.current.total).toBe(6000);

    act(() => {
      // Forward 3 more seconds
      now.mockReturnValue(countdownDate - 1000);
      jest.advanceTimersByTime(3000);
    });

    expect(onTick.mock.calls.length).toBe(9);
    expect(result.current.total).toBe(1000);

    act(() => {
      // The End: onComplete callback gets triggered instead of onTick
      now.mockReturnValue(countdownDate);
      jest.advanceTimersByTime(1000);
    });

    expect(onTick.mock.calls.length).toBe(9);
    expect(onTick).toBeCalledWith({
      ...defaultStats,
      total: 1000,
      seconds: 1,
    });

    expect(onComplete.mock.calls.length).toBe(1);
    expect(onComplete).toBeCalledWith({ ...defaultStats, completed: true }, false);
    expect(result.current.completed).toBe(true);
  });

  afterEach(async () => {
    await cleanup();
    jest.clearAllMocks();
  });
});
