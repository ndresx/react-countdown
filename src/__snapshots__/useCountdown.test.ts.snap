// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCountdown should use countdown 1`] = `
{
  "api": {
    "isCompleted": [Function],
    "isPaused": [Function],
    "isStarted": [Function],
    "isStopped": [Function],
    "pause": [Function],
    "start": [Function],
    "stop": [Function],
  },
  "completed": false,
  "days": 0,
  "formatted": {
    "days": "00",
    "hours": "00",
    "minutes": "00",
    "seconds": "10",
  },
  "hours": 0,
  "milliseconds": 0,
  "minutes": 0,
  "props": {
    "autoStart": true,
    "controlled": false,
    "date": 1482363377071,
    "daysInHours": false,
    "intervalDelay": 1000,
    "precision": 0,
    "pure": true,
    "zeroPadTime": 2,
  },
  "seconds": 10,
  "total": 10000,
}
`;
